build: restore stdmalloc builds

benesch opened the pull request:
stdmalloc builds were broken in the transition to using the C/C++ dependencies'
native build systems.

benesch committed:
build: restore stdmalloc builds

These were broken in the transition to using the C/C++ dependencies'
native build systems.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15719)
<!-- Reviewable:end -->

tamird commented:
Nice. Couple comments.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/common.mk, line 378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvSSsDBi184oIYmPa:-KjOvSSsDBi184oIYmPb:b-a5ez7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/build/common.mk#L378)):*
> ```Makefile
> 	  $(if $(findstring release,$(TYPE)),,-DWITH_$(if $(findstring mingw,$(TARGET_TRIPLE)),AVX2,SSE42)=OFF) \
> 	  -DSNAPPY_LIBRARIES=$(SNAPPY_DIR)/.libs/libsnappy.a -DSNAPPY_INCLUDE_DIR=$(SNAPPY_SRC_DIR) -DWITH_SNAPPY=ON \
> 	  $(if $(findstring stdmalloc,$(TAGS)),,-DJEMALLOC_LIBRARIES=$(JEMALLOC_DIR)/lib/libjemalloc.a -DJEMALLOC_INCLUDE_DIR=$(JEMALLOC_DIR)/include -DWITH_JEMALLOC=ON)
> ```

I think we want to put this thing in a whole separate directory if we're building stdmalloc, otherwise we're just busting caches each time.

---

*[pkg/ccl/storageccl/engineccl/jemalloc.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvx32XmoEBbFQL-nG:-KjOvx32XmoEBbFQL-nH:b1mtghq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/pkg/ccl/storageccl/engineccl/jemalloc.go#L21)):*
> ```Go
> package engineccl
> 
> // #cgo CPPFLAGS: -DJEMALLOC_NO_DEMANGLE
> ```

What does this do, here?

---

*[pkg/storage/engine/jemalloc.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOviBAIq0gjarwdhlg:-KjOviBAIq0gjarwdhlh:b1mtghq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/pkg/storage/engine/jemalloc.go#L21)):*
> ```Go
> package engine
> 
> // #cgo CPPFLAGS: -DJEMALLOC_NO_DEMANGLE
> ```

What does this do, here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-:-KjOvy1BnNyy96NkEtLp:b-wbhfma)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/common.mk, line 378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvSSsDBi184oIYmPa:-KjOwFiMFERfXPx1spx8:b-80fcno) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/build/common.mk#L378)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think we want to put this thing in a whole separate directory if we're building stdmalloc, otherwise we're just busting caches each time.
</blockquote></details>

It's worse than that: we're not even busting caches. how2fix? `BUILD_DIR := $(TARGET_TRIPLE)$(subst $(space),_,$(sort $(TAGS))`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[build/common.mk, line 378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvSSsDBi184oIYmPa:-KjOwpMJRnnAIDNShCXh:b-fjcdw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/build/common.mk#L378)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

It's worse than that: we're not even busting caches. how2fix? `BUILD_DIR := $(TARGET_TRIPLE)$(subst $(space),_,$(sort $(TAGS))`
</blockquote></details>

Hm, that would work, but it would make for an explosion of caches mostly unnecessarily. How about `ROCKSDB_DIR := $(if $(ISJEMALLOC),foo,bar)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[build/common.mk, line 378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvSSsDBi184oIYmPa:-KjOx6a143cFHHLM3Ecl:bemr6ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/build/common.mk#L378)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Hm, that would work, but it would make for an explosion of caches mostly unnecessarily. How about `ROCKSDB_DIR := $(if $(ISJEMALLOC),foo,bar)`?
</blockquote></details>

mk! On it.

---

*[pkg/ccl/storageccl/engineccl/jemalloc.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvx32XmoEBbFQL-nG:-KjOx--A2SWtneNL3z9G:b-6nzs56) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/pkg/ccl/storageccl/engineccl/jemalloc.go#L21)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What does this do, here?
</blockquote></details>

Nothing. Over-aggressive cargo-culting. Fixed.

---

*[pkg/storage/engine/jemalloc.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOviBAIq0gjarwdhlg:-KjOx1NBHFulXJGNoC_j:bnj3qhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/pkg/storage/engine/jemalloc.go#L21)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What does this do, here?
</blockquote></details>

Ditto. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/common.mk, line 378 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjOvSSsDBi184oIYmPa:-KjP3kQTpZqORX7fdrmB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e3a31f629fb6b7488c758dae9d13901147c2b5/build/common.mk#L378)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

mk! On it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/common.mk, line 261 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjP8UdxURd9gul9jVuP:-KjP8UdyyaNxWwv_J9Tc:b-ph9am4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a11583503e7e59009748bc49f6c5ac4ed0cc0bb/build/common.mk#L261)):*
> ```Makefile
> JEMALLOC_DIR := $(BUILD_DIR)/jemalloc
> PROTOBUF_DIR := $(BUILD_DIR)/protobuf
> ROCKSDB_DIR  := $(BUILD_DIR)/rocksdb$(if $(USE_STDMALLOC),_stdmalloc)
> ```

hm, should we invert this? seems more natural to have rocksdb and rocksdb_jemalloc. good thing we have rocksdb-rebuild!

i guess we already went the route of making std exceptional. either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-:-KjP8xZSsjV7RDKG8jln:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/common.mk, line 261 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719#-KjP8UdxURd9gul9jVuP:-KjQsNgwNKP6OfondveT:bef3sa9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a11583503e7e59009748bc49f6c5ac4ed0cc0bb/build/common.mk#L261)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, should we invert this? seems more natural to have rocksdb and rocksdb_jemalloc. good thing we have rocksdb-rebuild!

i guess we already went the route of making std exceptional. either way.
</blockquote></details>

Agreed, but since the Go tags are default-jemalloc, I figured this made more sense. I'd LG a PR to flip everything (and perhaps do it myself next week), but want to get this merged for now!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15719)*
<!-- Sent from Reviewable.io -->
